src/Components/Components/src/RenderTree/ArrayBuilder.cs

```diff
@@ -95,34 +96,51 @@ internal int Append(T[] source, int startIndex, int length)
         /// <param name="value">The value.</param>
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public void Overwrite(int index, in T value)
-            => _items[index] = value;
+        {
+            if (index > _itemsInUse)
```

Comment:
Shouldn't this be `if (index >= _itemsInUse)`? If `_itemsInUse == 0`, then this should throw even if `index` is `0`.

Thanks for adding the new unit tests for `ArrayBuilder`. Looks like we could use one to cover the boundary case for `Overwrite` when it should throw.